posts/2022/02/making-gpus-accessible-to-pydata-ecosystem-via-array-api.md

```diff
@@ -0,0 +1,520 @@
+have a direct equivalent in the Array API standard. This was mostly due to the
+fact that those functions are not generic enough to be defined in the Array
+API. Some examples of unimplemented functions that we encountered during this exercise:
+
+* `np.atleast_3d`
+* `np.strides`
+* `np.real`
+* `np.clip`
+
+
+#### 3. Inconsistencies between the NumPy API and the Array API standard
+
+Although the standard has managed to get good coverage of the NumPy API, there
+are places where there are inconsistencies between the two. This may be either
+function behaviour that is missing or different, or function or parameter names
+don't match (e.g., `concat` in the standard instead of `np.concatenate`).
+
+Here is an example of a naming inconsistency between NumPy and the Array API standard:
+
+```ipython
+
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Numpy API
+In [3]: np.concatenate
+Out[3]: <function numpy.concatenate>
+
+# Array API
+In [4]: npx.concat
+Out[4]: <function numpy.array_api._manipulation_functions.concat...>
+```
+
+And here is an example of a behavioral inconsistency for an indexing operation:
+
+```ipython
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Works with NumPy API
+In [3]: x = np.arange(10)
+In [4]: x[np.arange(2)]
+Out[4]: array([0, 1])
+
+
+# Doesn't Works with Array API
+In [5]: z = npx.arange(10)
+In [6]: z[npx.arange(2)]
+-----------------------------------------------------------------
+...
+IndexError: Non-zero dimensional integer array indices are not allowed in the array API namespace
+```
+
+Each of the inconsistencies can be worked around. Function renames may require some
+detective work by referring to the Array API docs, but are typically easy code changes.
+Differences in behavior can be harder to deal with at times. The Array API standard
+aims to provide a set of functionality that's orthogonal but complete - so if it's
+possible to perform some array operation on accelerator devices, then there's typically
+a way to access it in the standard."
+
+## Results
+
+After dealing with all the challenges above, our demo works with CuPy arrays as
+input. In this section, we will talk about about its performance on GPU vs.
+CPU.
+
+We ran this on AMD GPUs as well as NVIDIA GPUs. Below we show plots comparing
+the runtime of the demo code on CPU and GPU as a function of input array size.
+The x-axis of the plot is the image dimension, i.e the dimension of the original
+image after rescaling it by a constant factor along both dimensions. For this
+demo the rescaling factors were: 0.1, 0.2, 0.4, 0.6, 0.8 and 1.
+
+
+#### On AMD GPU:
+
+This was run on AMD Instinct MI50 32GB with 128 CPU cores.
+![NumPy cs CuPy AMD](/images/2022/02/numpy_vs_cupy_amd.png)
+
+
+#### On NVIDIA GPU:
+
+![NumPy cs CuPy NVIDIA](/images/2022/02/numpy_vs_cupy_nvidia.png)
+
+This was run on a NVIDIA TITAN RTX with 48 CPU cores. The GPU vs CPU plot is what you would
```

Comment:
Qgpu3 should have a Threadripper 3970X CPU, which is 32 physical (64 virtual) cores. Is 48 right here @aktech, @IvanYashchuk?

Reply:
NVIDIA TITAN RTX is on Qgpu2 and it has Threadripper 2970WX 24-Core (48 virtual) CPU. I don't know what machine was used for the plots.

Reply:
I used QGPU2 for this, hence 48 cores (counting virtual cores)

Reply:
Ah okay, thanks. Let's say "24 physical cores" then? I think that's the normal way to talk about N-core CPUs.

Reply:
I'd going to assume it's the same for the other CPU, and update that to "64 physical cores".